### PR TITLE
[WIP] 🏃‍♂️ run network policy conformance with antrea

### PR DESCRIPTION
**Pull request by ashish-amarnath**
Signed-off-by: Ashish Amarnath <ashisham@vmware.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Runs NetworkPolicy conformance tests using Antrea as the CNI plugin.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note

```

cc @dims

### CODE REVIEW

**Comment by ashish-amarnath** [comment]
If there is a way to template the e2e conf files, please lmk. atm it is a copy-paste of the e2e-conf.yaml with the only change being the file referring to the CNI plugin

File: Makefile [hunk 1]
```diff
@@ -36,8 +36,10 @@ TEST_E2E_DIR := test/e2e
 E2E_DATA_DIR ?= $(REPO_ROOT)/test/e2e/data
 E2E_CONF_PATH  ?= $(E2E_DATA_DIR)/e2e_conf.yaml
 E2E_EKS_CONF_PATH ?= $(E2E_DATA_DIR)/e2e_eks_conf.yaml
+E2E_CONF_CNI_ANTREA ?= $(E2E_DATA_DIR)/e2e_conf_cni_antrea.yaml
 KUBETEST_CONF_PATH ?= $(abspath $(E2E_DATA_DIR)/kubetest/conformance.yaml)
 KUBETEST_FAST_CONF_PATH ?= $(abspath $(E2E_DATA_DIR)/kubetest/conformance-fast.yaml)
+KUBETEST_NETWORKPOLICY_CONF_PATH ?= $(abspath $(E2E_DATA_DIR)/kubetest/conformance-networkpoicy.yaml)
```
Comment (dims) [inline Makefile:42]: typo ... `networkpoicy`

File: test/e2e/data/kubetest/conformance-networkpoicy.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+ ## TODO(ashish-amarnath): get the correct focus string to run network policy tests
+ginkgo.focus: \[Conformance\]
```
Comment (dims) [inline test/e2e/data/kubetest/conformance-networkpoicy.yaml:2]: we should have a CI job for conformance and a separate one for the network policy as well.

Comment (ashish-amarnath) [reply test/e2e/data/kubetest/conformance-networkpoicy.yaml:2]: Yes, I agree. I didn't know the focus expression for those tests. so just leaving it like this for now till I find out what the right focus expression is.

Comment (dims) [reply test/e2e/data/kubetest/conformance-networkpoicy.yaml:2]: @jayunit100 will know for sure!

File: Makefile (hunk 1 above)
Comment (dims) [reply Makefile:42]: line 42 in this yaml file is still `poicy`

Comment (ashish-amarnath) [reply Makefile:42]: opps.. didn't save it

**Comment by ashish-amarnath** [comment]
cc @antoninbas

File: test/e2e/data/cni/antrea.yaml [hunk 3]
```diff
@@ -0,0 +1,480 @@
+---
+# source: https://github.com/vmware-tanzu/antrea/releases/download/v0.1.0/antrea.yml
```
Comment (antoninbas) [inline test/e2e/data/cni/antrea.yaml:2]: using this version definitely does not seem right :)
This is the first version of Antrea (from 2019), and the latest version is v0.13.1

is it possible to check-in a version number in a file, and download the YAML automatically? Or does the whole YAML need to be checked-in and updated each time?

File: Makefile [hunk 4]
```diff
@@ -169,6 +171,12 @@ test-conformance: $(GINKGO) $(KIND) $(SSM_PLUGIN) $(KUSTOMIZE) e2e-image ## Run
 
 test-conformance-fast: ## Run clusterctl based conformance test on workload cluster (requires Docker) using a subset of the conformance suite in parallel.
 	$(MAKE) test-conformance CONFORMANCE_E2E_ARGS="-kubetest.config-file=$(KUBETEST_FAST_CONF_PATH) -kubetest.ginkgo-nodes=5 $(E2E_ARGS)"
+
+# TODO(ashish-amarnath) what does "kubetest.ginkgo-nodes=5" do? is it required here?
+.PHONY: test-conformance-cni-antrea
+test-conformance-cni-antrea: $(GINKGO) $(KIND) $(SSM_PLUGIN) $(KUSTOMIZE) e2e-image ## Run network policy conformance tests on workload cluster using Antrea as CNI
```
Comment (antoninbas) [inline Makefile:177]: any thoughts on making this more generic, as described in https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1745? cc @dims

File: test/e2e/data/cni/antrea.yaml (hunk 3 above)
Comment (ashish-amarnath) [reply test/e2e/data/cni/antrea.yaml]: I think the thing responsible for applying CNI is prepared to receive yaml that can be applied to the cluster.
I've updated to use the v0.13.1 yaml

File: test/e2e/data/e2e_conf_cni_antrea.yaml [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        targetName: "cluster-template-spot-instances.yaml"
+      - sourcePath: "./infrastructure-aws/cluster-template-mhc.yaml"
+        targetName: "cluster-template-mhc.yaml"
+      - sourcePath: "./infrastructure-aws/cluster-template-machine-pool.yaml"
+        targetName: "cluster-template-machine-pool.yaml"
+      - sourcePath: "./infrastructure-aws/cluster-template-upgrade-to-main.yaml"
+        targetName: "cluster-template-upgrade-to-main.yaml"
+variables:
+  KUBERNETES_VERSION: "v1.20.4"
+  CNI: "../../data/cni/antrea.yaml"
```
Comment (sedefsavas) [inline test/e2e/data/e2e_conf_cni_antrea.yaml:116]: No need to have a separate config to use a different CNI.

CAPZ has an example of testing different CNIs with single config here:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/2d58abba334517ad59a3f2c7f6d8dd476bfb0f49/test/e2e/config/azure-dev.yaml#L85-L87

This is how they are embedding CNI resources into a template:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/4ba3dc99f2f12cd58cc3bb332b39cd9db91e1633/test/e2e/e2e_suite_test.go#L351

Comment (ashish-amarnath) [reply test/e2e/data/e2e_conf_cni_antrea.yaml:116]: Thanks @sedefsavas I'll look at this.

**Comment by sedefsavas** [comment]
I think we can have a seperate conformance test under conformance suite that uses a new template that has antrea as CNI.

```
conformance/
    conformance_test.go
    conformance_test_with_antrea.go
```

File: Makefile (hunk 4 above)
Comment (ashish-amarnath) [reply Makefile:177]: Yep.. looks like we have an example from the capz repo. I'll look into it.

**Comment by ashish-amarnath** [comment]
I haven't had a chance to get back to this from earlier. I will try to make time for this next week.

**Comment by randomvariable** [comment]
We've discussed this offline. 

CAPA maintainers (particularly non-VMware) are not in a position to become a testing point and CI signal maintainers for multiple CNIs.

The path forward here is to take https://github.com/kubernetes-sigs/cluster-api/pull/4041 to completion. This will then let you consume Cluster API AWS in the Antrea repo and have CI signal testing against the main branch of Kubernetes.

**Pull request closed without merging**